core/server/common/src/main/java/alluxio/master/journal/raft/JournalStateMachine.java

```diff
@@ -70,35 +88,164 @@
   private volatile long mNextSequenceNumberToRead = 0;
   private volatile boolean mSnapshotting = false;
+  private volatile boolean mIsLeader = false;
+
   // The start time of the most recent snapshot
   private volatile long mLastSnapshotStartTime = 0;
   /** Used to control applying to masters. */
   private BufferedJournalApplier mJournalApplier;
+  private final SimpleStateMachineStorage mStorage = new SimpleStateMachineStorage();
+  private RaftGroupId mRaftGroupId;
 
   /**
-   * @param journals master journals; these journals are still owned by the caller, not by the
-   *        journal state machine
-   * @param journalSinks a supplier for journal sinks
+   * @param journals     master journals; these journals are still owned by the caller, not by the
+   *                     journal state machine
+   * @param journalSystem the raft journal system
    */
-  public JournalStateMachine(Map<String, RaftJournal> journals,
-      Supplier<Set<JournalSink>> journalSinks) {
+  public JournalStateMachine(Map<String, RaftJournal> journals, RaftJournalSystem journalSystem) {
     mJournals = journals;
-    mJournalApplier = new BufferedJournalApplier(journals, journalSinks);
+    mJournalApplier = new BufferedJournalApplier(journals,
+        () -> journalSystem.getJournalSinks(null));
     resetState();
     LOG.info("Initialized new journal state machine");
+    mJournalSystem = journalSystem;
+  }
+
+  private static <T> CompletableFuture<T> completeExceptionally(Exception e) {
+    final CompletableFuture<T> future = new CompletableFuture<>();
+    future.completeExceptionally(e);
+    return future;
+  }
+
+  @Override
+  public void initialize(RaftServer server, RaftGroupId groupId,
+      RaftStorage raftStorage) throws IOException {
+    getLifeCycle().startAndTransition(() -> {
+      super.initialize(server, groupId, raftStorage);
+      mRaftGroupId = groupId;
+      mStorage.init(raftStorage);
+      loadSnapshot(mStorage.getLatestSnapshot());
+    });
+  }
+
+  @Override
+  public void reinitialize() throws IOException {
+    LOG.info("Reinitializing state machine.");
+    mStorage.loadLatestSnapshot();
+    loadSnapshot(mStorage.getLatestSnapshot());
+    unpause();
+  }
+
+  private void loadSnapshot(SingleFileSnapshotInfo snapshot) throws IOException {
+    if (snapshot == null) {
+      LOG.info("No snapshot to load");
+      return;
+    }
+    LOG.info("Loading Snapshot {}", snapshot);
+    final File snapshotFile = snapshot.getFile().getPath().toFile();
+    if (!snapshotFile.exists()) {
+      LOG.error("The snapshot {} does not exist", snapshot);
+      throw new FileNotFoundException(
+          String.format("The snapshot file %s does not exist", snapshotFile.getPath()));
+    }
+    try {
+      resetState();
+      setLastAppliedTermIndex(snapshot.getTermIndex());
+      install(snapshotFile);
+    } catch (Exception e) {
+      LOG.error("Failed to load snapshot {}", snapshot, e);
+      throw e;
```

Comment:
Should we throw a new exception with the message like above? Also, do we need to log and throw an exception?

Reply:
Removed the logging message and included it in a new exception.